cpp/src/toppra/solver/qpOASES-wrapper.cpp

```diff
@@ -0,0 +1,127 @@
+#include <toppra/solver/qpOASES-wrapper.hpp>
+
+#include <qpOASES.hpp>
+
+namespace toppra {
+namespace solver {
+struct qpOASESWrapper::Impl {
+  qpOASES::SQProblem qp;
+
+  Impl(Eigen::Index nV, Eigen::Index nC)
+    : qp (nV, nC)
+  {
+    qpOASES::Options options;
+    options.printLevel = qpOASES::PL_NONE;
+
+    qp.setOptions( options );
+  }
+};
+
+qpOASESWrapper::qpOASESWrapper (const LinearConstraintPtrs& constraints, const GeometricPath& path,
+        const Vector& times)
+  : Solver (constraints, path, times)
+{
+  // Currently only support Canonical Linear Constraint
+  Eigen::Index nC = 2; // First constraint is x + 2 D u <= xnext_max, second is xnext_min <= x + 2D u
+  for (const Solver::LinearConstraintParams& linParam : constraintsParams_.lin)
+    nC += linParam.F[0].rows();
+
+  Eigen::Index nV (nbVars());
+  assert(nV == 2);
+  A_  = RMatrix::Zero(nC, nV);
+  lA_ = -Vector::Ones(nC);
+  hA_ = -Vector::Ones(nC);
+
+  impl_ = new Impl(nV, nC);
```

Comment:
Why are you not using a smart pointer here?

Reply:
`impl` is not accessible from outside the class. It is meant to avoid exposing qpOASES so that qpOASES is an internal dependency and not an external one. It is based on the opaque pointer pattern (https://en.wikipedia.org/wiki/Opaque_pointer)

I will change it to a `unique_ptr`.

Reply:
Note that a raw pointer here is fairly safe. The risk of memory leak is very limited (if it exists at all).